Test/tutorial/Popit.dfy

```diff
@@ -0,0 +1,963 @@
+    case 5 => "5" case 6 => "6" case 7 => "7" case 8 => "8" case 9 => "9"
+    case _ => natToString(n / 10) + natToString(n % 10)
+  }
+
+/// We write the second function to convert a string-encoded natural number to a nat
+/// Note the following interesting thing: We recurse on the |s|-1 initial elements of the strings,
+/// Recursing that way makes proof to be straightforward for Dafny.
+/// The alternative would have been an auxiliary function to accumulate intermediate results,
+/// but the proofs become complicated if not unfeasible
+
+  function stringToNat(s: stringNat): nat
+    decreases |s|
+  {
+    if |s| == 1 then
+      match s[0]
+      case '0' => 0 case '1' => 1 case '2' => 2 case '3' => 3 case '4' => 4
+      case '5' => 5 case '6' => 6 case '7' => 7 case '8' => 8 case '9' => 9
+    else
+      stringToNat(s[..|s|-1])*10 + stringToNat(s[|s|-1..|s|])
+  }
+
+/// Our functions are so well defined that Dafny is able to automatically prove
+/// the two equivalence lemmas
+
+  lemma natToStringThenStringToNatIdem(n: nat)
+    ensures stringToNat(natToString(n)) == n
+  { // Proof is automatic
+  }
+  lemma stringToNatThenNatToStringIdem(s: stringNat)
+    ensures natToString(stringToNat(s)) == s
+  { // Proof is automatic
+  }
+}
+
+/// Ok, now let's move to the really interesting part
+/// We wrap everything in a module to be able to use the new function syntax
+/// That way, "function" and "predicate" are compilable by default
+/// We also use the new quantifier syntax so that the following expression
+/// does not result in a parser error. Otherwise, it would try to interpret this as a tuple and would fail.
+/// assert (forall x <- popIt | IsMin(popIt, x), y <- popIt | IsMin(popIt, y) :: x == y);
+
+module {:options "/functionSyntax:4", "/quantifierSyntax:4"} PopItWorld {
+  import Printer
+
+/// What is a pop-it game really? Rows with bubbles? 
+/// We could encode it in several ways: sequence of sequences of bubbles that are
+/// popped or not
+/// Or we could just store sequence of sequences of unpopped bubbles.
+/// Here we choose the multiset encoding of a PopIt game. This means each
+/// number n in the multiset represents a row with a number of n consecutives bubbles.
+/// So a PopIt game is a multiset of non-negative integers.
+/// Let us define it with a predicate.
+/// Since this predicate is never going to be executed, we can mark it as ghost
+
+  ghost predicate IsPopit(x: multiset<int>) {
+    forall row <- x :: row > 0
+  }
+
+/// Now we define a PopIt as a subset type of a multiset<int> as expected:
+/// We decide that the empty multi set is a winning PopIt game,
+/// so we don't need to provide a witness since multiset{} is a valid PopIt game
+
+  type PopIt = x: multiset<int> | IsPopit(x)
+
+/// Ok, let's define one useful predicate. A PopIt is said final if it's no longer possible to play
+
+  predicate IsFinal(x: PopIt) {
+    |x| == 0
+  }
```

Comment:
It would be nice to make this a member function of a datatype, if it were possible